### ISSUE TITLE
SocketException after copying many photos

### ISSUE THREAD

**Issue by gmmapowell** [vidstige/jadb#109]
I'm trying to use JADB to automatically offload photos from my (various) Android devices.  I've found it very easy to use and have got up and running in an hour or so.

But I'm encountering a socket exception about every 250 or so photos that I pull across; I'm guessing that something is incapable of coping with 800+MB of file transfer (probably failing to clear something from memory), but I'm not really sure.  The pull method returns void, so it doesn't seem like I am failing to clean anything up myself, and I don't see any cleanup method in the RemoteFile structure.

I tried googling for anybody else with this problem but didn't see anything.

The exact exception I'm seeing is:

```
java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:189)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at java.io.DataInputStream.readFully(DataInputStream.java:195)
	at java.io.DataInputStream.readFully(DataInputStream.java:169)
	at se.vidstige.jadb.Transport.readString(Transport.java:46)
	at se.vidstige.jadb.Transport.verifyResponse(Transport.java:36)
	at se.vidstige.jadb.JadbDevice.getTransport(JadbDevice.java:52)
	at se.vidstige.jadb.JadbDevice.pull(JadbDevice.java:183)
	at se.vidstige.jadb.JadbDevice.pull(JadbDevice.java:192)
	at com.gmmapowell.mediabackup.Main.backupAdb(Main.java:115)
	at com.gmmapowell.mediabackup.Main.main(Main.java:62)
```

I'm using v1.1.0 from the jitstack maven repo.

**Comment by vidstige** [comment]
Thanks for reporting this bug. That's curious. It should not be due to memory problems, but one can never be sure. The exception happens when verifying the response from the device. `Connection reset` typically means the remote host (device) closed the connection. Could you post your source code as well?

**Comment by gmmapowell** [comment]
The actual code I was using was embedded in a much bigger application, so I've extracted the key elements below.

```
import java.io.File;
import java.util.List;

import se.vidstige.jadb.JadbConnection;
import se.vidstige.jadb.JadbDevice;
import se.vidstige.jadb.RemoteFile;

public class BugReport {
	private static File tmp;

	public static void main(String[] args) throws Exception {
		tmp = File.createTempFile("img", ".jpg");
		tmp.deleteOnExit();

		JadbDevice dadb = connect();
		String fromdir = "/sdcard/DCIM/100ANDRO"; // this is device-dependent
		for (RemoteFile rf : dadb.list(fromdir)) {
			if (rf.getPath().startsWith("."))
				continue;
			readFile(dadb, fromdir, rf);
		}
		tmp.delete();
	}
	
	private static JadbDevice connect() throws Exception {
		JadbConnection conn = new JadbConnection();
		List<JadbDevice> devices = conn.getDevices();
		if (devices.isEmpty())
			throw new RuntimeException("No devices found");
		return devices.get(0);
	}

	private static void readFile(JadbDevice dadb, String fromdir, RemoteFile rf) throws Exception{
		// This feels wrong to me, but this is how it seems to work
		// getName() throws an exception
		rf = new RemoteFile(fromdir + "/" + rf.getPath());
		System.out.println(rf.getPath());
		dadb.pull(rf, tmp);
	}
}
```
And this is the output I get:

```
DSC_0001.JPG
DSC_0002.JPG
DSC_0051.JPG
DSC_0003.JPG
DSC_0012.JPG
DSC_0004.JPG
DSC_0005.JPG
DSC_0006.JPG
DSC_0007.JPG
DSC_0008.JPG
DSC_0009.JPG
DSC_0010.JPG
DSC_0011.JPG
DSC_0021.JPG
DSC_0013.JPG
DSC_0014.JPG
DSC_0015.JPG
DSC_0016.JPG
DSC_0017.JPG
DSC_0018.JPG
DSC_0019.JPG
DSC_0020.JPG
DSC_0024.JPG
DSC_0022.JPG
DSC_0023.JPG
DSC_0027.JPG
DSC_0025.JPG
DSC_0026.JPG
DSC_0046.JPG
DSC_0028.JPG
DSC_0029.JPG
DSC_0030.JPG
DSC_0031.JPG
DSC_0032.JPG
DSC_0033.JPG
DSC_0034.JPG
DSC_0035.JPG
DSC_0036.JPG
DSC_0037.JPG
DSC_0038.JPG
DSC_0039.JPG
DSC_0040.JPG
DSC_0041.JPG
DSC_0042.JPG
DSC_0043.JPG
DSC_0044.JPG
DSC_0045.JPG
DSC_0047.JPG
DSC_0048.JPG
DSC_0079.JPG
DSC_0049.JPG
DSC_0050.JPG
DSC_0052.JPG
DSC_0053.JPG
DSC_0054.JPG
DSC_0055.JPG
DSC_0056.JPG
DSC_0059.JPG
DSC_0057.JPG
DSC_0058.JPG
DSC_0061.JPG
DSC_0060.JPG
DSC_0063.JPG
DSC_0062.JPG
DSC_0064.JPG
DSC_0065.JPG
DSC_0066.JPG
DSC_0067.JPG
DSC_0068.JPG
DSC_0069.JPG
DSC_0070.JPG
DSC_0071.JPG
DSC_0072.JPG
DSC_0073.JPG
DSC_0074.JPG
DSC_0075.JPG
DSC_0076.JPG
DSC_0077.JPG
DSC_0078.JPG
DSC_0080.JPG
DSC_0081.JPG
DSC_0082.JPG
DSC_0083.JPG
DSC_0084.JPG
DSC_0085.JPG
DSC_0086.JPG
DSC_0087.JPG
DSC_0089.JPG
DSC_0090.JPG
DSC_0091.JPG
DSC_0092.JPG
DSC_0088.JPG
DSC_0093.JPG
DSC_0094.JPG
DSC_0095.JPG
DSC_0098.JPG
DSC_0097.JPG
DSC_0099.JPG
DSC_0100.JPG
DSC_0101.JPG
DSC_0102.JPG
DSC_0103.JPG
DSC_0104.JPG
DSC_0105.JPG
DSC_0106.JPG
DSC_0107.JPG
DSC_0108.JPG
DSC_0109.JPG
MOV_0110.mp4
MOV_0111.mp4
DSC_0112.JPG
DSC_0113.JPG
DSC_0114.JPG
DSC_0115.JPG
DSC_0116.JPG
DSC_0117.JPG
DSC_0118.JPG
DSC_0119.JPG
DSC_0120.JPG
DSC_0121.JPG
DSC_0122.JPG
DSC_0123.JPG
DSC_0124.JPG
DSC_0125.JPG
DSC_0126.JPG
DSC_0127.JPG
DSC_0128.JPG
DSC_0129.JPG
DSC_0130.JPG
DSC_0131.JPG
DSC_0132.JPG
DSC_0133.JPG
DSC_0134.JPG
DSC_0135.JPG
DSC_0136.JPG
DSC_0137.JPG
DSC_0138.JPG
DSC_0139.JPG
DSC_0140.JPG
DSC_0141.JPG
DSC_0142.JPG
DSC_0143.JPG
DSC_0144.JPG
DSC_0145.JPG
DSC_0146.JPG
DSC_0147.JPG
DSC_0148.JPG
DSC_0149.JPG
DSC_0150.JPG
DSC_0151.JPG
DSC_0152.JPG
DSC_0153.JPG
DSC_0154.JPG
DSC_0155.JPG
DSC_0156.JPG
DSC_0157.JPG
DSC_0158.JPG
DSC_0159.JPG
DSC_0160.JPG
DSC_0161.JPG
DSC_0162.JPG
DSC_0163.JPG
DSC_0164.JPG
DSC_0165.JPG
DSC_0166.JPG
DSC_0167.JPG
DSC_0168.JPG
DSC_0169.JPG
DSC_0170.JPG
DSC_0171.JPG
DSC_0172.JPG
MOV_0173.mp4
DSC_0174.JPG
DSC_0175.JPG
DSC_0176.JPG
DSC_0177.JPG
DSC_0178.JPG
DSC_0179.JPG
DSC_0180.JPG
DSC_0181.JPG
DSC_0182.JPG
DSC_0183.JPG
DSC_0184.JPG
DSC_0185.JPG
DSC_0186.JPG
DSC_0187.JPG
DSC_0188.JPG
DSC_0189.JPG
DSC_0190.JPG
DSC_0191.JPG
DSC_0192.JPG
DSC_0193.JPG
DSC_0194.JPG
DSC_0195.JPG
DSC_0196.JPG
DSC_0197.JPG
DSC_0198.JPG
DSC_0199.JPG
DSC_0200.JPG
DSC_0201.JPG
DSC_0202.JPG
DSC_0203.JPG
DSC_0204.JPG
DSC_0205.JPG
DSC_0206.JPG
DSC_0207.JPG
DSC_0208.JPG
DSC_0209.JPG
DSC_0210.JPG
DSC_0211.JPG
DSC_0212.JPG
DSC_0213.JPG
DSC_0214.JPG
DSC_0215.JPG
DSC_0216.JPG
DSC_0217.JPG
DSC_0218.JPG
DSC_0219.JPG
DSC_0220.JPG
DSC_0221.JPG
DSC_0222.JPG
DSC_0223.JPG
DSC_0224.JPG
DSC_0225.JPG
DSC_0226.JPG
DSC_0227.JPG
DSC_0228.JPG
DSC_0229.JPG
DSC_0230.JPG
DSC_0231.JPG
DSC_0232.JPG
DSC_0233.JPG
DSC_0234.JPG
DSC_0235.JPG
DSC_0236.JPG
DSC_0237.JPG
DSC_0238.JPG
DSC_0239.JPG
DSC_0240.JPG
DSC_0241.JPG
DSC_0242.JPG
DSC_0243.JPG
DSC_0244.JPG
DSC_0245.JPG
DSC_0246.JPG
DSC_0247.JPG
DSC_0248.JPG
Exception in thread "main" java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:189)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at java.io.DataInputStream.readFully(DataInputStream.java:195)
	at java.io.DataInputStream.readFully(DataInputStream.java:169)
	at se.vidstige.jadb.Transport.readString(Transport.java:46)
	at se.vidstige.jadb.Transport.verifyResponse(Transport.java:36)
	at se.vidstige.jadb.JadbDevice.getTransport(JadbDevice.java:52)
	at se.vidstige.jadb.JadbDevice.pull(JadbDevice.java:183)
	at se.vidstige.jadb.JadbDevice.pull(JadbDevice.java:192)
	at com.gmmapowell.mediabackup.BugReport.readFile(BugReport.java:38)
	at com.gmmapowell.mediabackup.BugReport.main(BugReport.java:22)
```

Is there any debugging/logging/tracing I can turn on, or anything I can do on the phone to get more data?

Note that I've removed all my attempts at exception handling.  With these turned on, the exception repeats about 20-30 times before finally resetting the connection, then it will do another 200-300 pulls.

**Comment by SKART1** [comment]
Cool usage of the library :+1: 

You may try to see the system log in the logcat (turn it to bee maximum size in developer options) 
It is stored for some time - so you may try to reconnect phone and ask `adb logcat`


Also as a quick fix you may try to copy files in batches closing and opening connection after each batch

**Comment by vidstige** [comment]
By a fortunate turn of events, there is a PR up that needs testing _and_ it might also solve this. In #95  sockets are systematically closed at the correct times. Please try that. I don't know what would be the easiest way to test it though. Perhaps just check out that branch locally?

**Comment by gmmapowell** [comment]
Fortunate, indeed.

Yes, that seems to work just fine.  I'm not sure what the differences were, but it ran through all 800+ files without a hitch.

**Comment by vidstige** [comment]
Wow, beautiful. Thank you for testing! :bowing_man: I'll merge it promptly

**Comment by vidstige** [comment]
I just merged #95 to master, so feel free to use the bleeding edge version (latest master) for now. I'll try to make a new release some time soonish. Very cool project, btw! :)
